Feature/1123 laplace approx

mitzimorris opened the pull request:
#### Submisison Checklist

- [ ] Run tests: `./runCmdStanTests.py src/test`
- [ ] Declare copyright holder and open-source license: see below

#### Summary:

#### Intended Effect:

Expose Stan service method `laplace_sample` to CmdStan interface, (https://github.com/stan-dev/stan/pull/3148).

The `laplace_sample` specific arguments are:

- `draws` - number of draws in sample
- `jacobian` - Boolean flag, True == 1 (default), False == 0
- `mode` - Either a JSON file containing estimates for all model parameters or a Stan CSV file produced by the `optimize` method

As part of this PR I added file `src/cmdstan/command_helper.hpp` for auxiliary helper methods, similar in spirit to `stansummary_helper.hpp`.     The goal is to improve the readability and maintainability `command.hpp`.   Further cleanup is needed, but that should be done in another PR.

#### How to Verify:

Unit test `laplace_test`

#### Side Effects:

N/A

#### Documentation:

Documentation will be added via a separate PR in the docs repo.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

mitzimorris commented:
Filed issue https://github.com/stan-dev/cmdstan/issues/1129 to continue code cleanup effort.

mitzimorris commented:
> A few small comments now. I think a full review should wait until after the feature freeze ends next week.

agreed.

> Question: should the argument to specify the Jacobians to be used for optimization be added simultaneously?

depends on how granular you like your PRs.   This could be done as a sub-branch from this branch, correct?

waiting until after feature freeze would allow for more time to do more on https://github.com/stan-dev/cmdstan/issues/1129

WardBrian commented:
> depends on how granular you like your PRs

Fair enough, my impression thus far has been that Laplace approximation is mostly useful if you were able to optimize with Jacobians. If it's useful on its own I don't think it is necessary to include in this PR or one before this

bob-carpenter commented on src/cmdstan/arguments/arg_laplace_jacobian.hpp:
```diff
@@ -0,0 +1,23 @@
+ public:
+  arg_laplace_jacobian() : bool_argument() {
+    _name = "jacobian";
+    _description = "Jacobian adjustment?";
```
this can just be "set to true to include change-of-variables adjustment for constraining parameter transforms"

I don't think we want to go into True corresponding to max a posteriori (MAP) and False corresponding to a (penalized) maximum likelihood estimate (MLE).

mitzimorris replied:
fixed.

bob-carpenter commented:
> Laplace approximation is mostly useful if you were able to optimize with Jacobians

Without Jacobian adjustments, optimization gives you the penalized MLE and the corresponding Laplace approximation gives you approximate standard errors.

With Jacobian adjustments, optimization gives you the max a posteriori (MAP) estimate and the corresponding Laplace approximation gives you an approximate posterior.

It's just a matter of how much we document the actual computation and how much we doc its statistical application.

bob-carpenter commented:
I should also note that the existing RStan implementation is the MLE plus standard errors version (no Jacobian adjustment).  So just to recreate that feature we need the Jacobian adjustment off option for the Laplace approximation.

mitzimorris commented on src/cmdstan/arguments/arg_laplace_draws.hpp:
```diff
@@ -0,0 +1,26 @@
+ public:
+  arg_laplace_draws() : int_argument() {
+    _name = "draws";
+    _description = "Numer of draws from the laplace apporximation";
```
good catch

mitzimorris commented on src/cmdstan/arguments/arg_laplace_mode.hpp:
```diff
@@ -0,0 +1,26 @@
+    _name = "mode";
+    _description
+        = "A specification of a mode on the constrained scale "
+          "for all model parameters";
```
added

mitzimorris commented:
made suggested changes.   will branch from this branch to work on issues #1124 and #1129

mitzimorris commented:
as part of thie PR, will remove duplicate "jacobian" argument classes, so that log_prob, optimize, and laplace all use same class.

WardBrian commented:
> not possible - method log_prob uses the more verbose name "jacobian_adjust". won't change log_prob. both optimize and laplace will use shorter arg name "jacobian", unless @bob-carpenter and/or @WardBrian would prefer name "jacobian_adjust".

I think we can change log_prob without much fear - it is not in a released version yet and shouldn't have many users. If you want to try to sneak that in before the release we could, see #1130

mitzimorris commented:
implementing #1124 as part of this PR - adding option for jacobian correction to optimization method.

for optimization, the default value of parameter "jacobian" is false.   to implement this, under the hood, defined new arg type "arg_jacobian_false" with the correct default.  this preserves backwards compatibility.

added test to check jacobians, and rewrote optimization tests, adding checks that specified algorithm is invoked.

mitzimorris commented:
ready for another round of review.

bob-carpenter commented:
@mitzimorris: It looks like there are conflicts that need to be sorted before a review.
